datalad/auto.py

```diff
@@ -0,0 +1,184 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Proxy basic file operations (such as open) to obtain files automagically upon I/O
+"""
+
+from mock import patch
+from six import PY2
+import six.moves.builtins as __builtin__
+builtins_name = '__builtin__' if PY2 else 'builtins'
+
+try:
+    import h5py
+except ImportError:
+    h5py = None
+
+import logging
+
+from os.path import dirname, abspath, pardir, join as opj, exists, basename
+
+from .support.annexrepo import AnnexRepo
+from .support.gitrepo import GitRepo
+from .support.exceptions import CommandError
+from .cmd import Runner
+
+from .utils import swallow_outputs
+lgr = logging.getLogger("datalad.auto")
+
+
+# TODO: shouldn't be some classmethod of GitRepo?
```

Comment:
Hm. May be worth measuring whether a ls-files call is actually faster than checking indexed_files.

Regarding the "read-only" mode: Any git-annex call will execute an init if it wasn't done before, as far as I'm aware. So, what exactly do you expect to be different in that mode?